src/utilities.cpp

```diff
@@ -559,6 +562,347 @@ void findAllVariablesWithEquivalences(const ComponentPtr &component, VariablePtr
+{
+    XmlNodePtr childNode = node->firstChild();
+    while (childNode != nullptr) {
+        if (childNode->isMathmlElement("cn")) {
+            std::string unitsName = childNode->attribute("units");
+            auto foundNameIter = replaceMap.find(unitsName);
+            if (foundNameIter != replaceMap.end()) {
+                childNode->setAttribute("units", foundNameIter->second.c_str());
+            }
+        }
+        findAndReplaceCnUnitsNames(childNode, replaceMap);
+        childNode = childNode->next();
+    }
+}
+
+void findAndReplaceComponentCnUnitsNames(const ComponentPtr &component, const StringStringMap &replaceMap)
+{
+    std::string mathContent = component->math();
+    if (mathContent.empty()) {
+        return;
+    }
+    bool contentModified = false;
+    std::string newMathContent;
+    std::vector<XmlDocPtr> mathDocs = multiRootXml(mathContent);
+    for (const auto &doc : mathDocs) {
+        auto rootNode = doc->rootNode();
+        if (rootNode->isMathmlElement("math")) {
+            auto originalMath = rootNode->convertToString();
+            findAndReplaceCnUnitsNames(rootNode, replaceMap);
+            auto newMath = rootNode->convertToString();
+            newMathContent += newMath;
+            if (newMath != originalMath) {
+                contentModified = true;
+            }
+        }
+    }
+
+    if (contentModified) {
+        component->setMath(newMathContent);
+    }
+}
+
+void findAndReplaceComponentsCnUnitsNames(const ComponentPtr &component, const StringStringMap &replaceMap)
+{
+    findAndReplaceComponentCnUnitsNames(component, replaceMap);
+    for (size_t index = 0; index < component->componentCount(); ++index) {
+        auto childComponent = component->component(index);
+        findAndReplaceComponentCnUnitsNames(childComponent, replaceMap);
+    }
+}
+
+size_t getComponentIndexInComponentEntity(const ComponentEntityPtr &componentParent, const ComponentEntityPtr &component)
+{
+    size_t index = 0;
+    bool found = false;
+    while (index < componentParent->componentCount() && !found) {
+        if (componentParent->component(index) == component) {
+            found = true;
+        } else {
+            ++index;
+        }
+    }
+
+    return index;
+}
+
+IndexStack reverseEngineerIndexStack(const ComponentPtr &component)
+{
+    auto dummyVariable = Variable::create();
+    component->addVariable(dummyVariable);
+    IndexStack indexStack = reverseEngineerIndexStack(dummyVariable);
+    indexStack.pop_back();
+    component->removeVariable(dummyVariable);
+
+    return indexStack;
+}
+
+IndexStack rebaseIndexStack(const IndexStack &stack, const IndexStack &originStack, const IndexStack &destinationStack)
+{
+    auto rebasedStack = stack;
+
+    rebasedStack.resize(originStack.size(), SIZE_MAX);
+    if (rebasedStack == originStack) {
+        rebasedStack = destinationStack;
+        auto offsetIt = stack.begin() + static_cast<int64_t>(originStack.size());
```

Comment:
Replace with:
```
auto offsetIt = stack.begin() + int64_t(originStack.size());
```

Reply:
Hmm ... clang used to complain about "old style C casts" like that.  I thought it was discouraged, so what's the rationale for wanting this to change?

Reply:
I suggested a C++ style cast, which makes it easier to read. FWIW, a C style cast would have been:
```
auto offsetIt = stack.begin() + (int64_t) originStack.size();
```
and I agree that a C style cast is discouraged.

Reply:
5584669c